### PR TITLE
[PyTorch][3/N] Enable the rest forward spec options for ShardedEmbedding and ShardedEmbeddingBag.

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68096
* #68786
* #67705
* __->__ #67799

We have enabled the sharding embedding and embedding bag in https://github.com/pytorch/pytorch/pull/67188 and https://github.com/pytorch/pytorch/pull/66604. We now want to enable as many parameters as defined in doc as possible: https://pytorch.org/docs/stable/generated/torch.nn.functional.embedding_bag.html, https://pytorch.org/docs/stable/generated/torch.nn.functional.embedding.html.

For the ones that we don't support we just throw exception.

Last but not least, we use get to get params instead of directly using the key.

Differential Revision: [D31985333](https://our.internmc.facebook.com/intern/diff/D31985333/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

File: torch/distributed/_sharded_tensor/ops/_common.py [hunk 1]
```diff
@@ -268,3 +310,51 @@ def _communicate_list_to_each_rank(
...
+    pg,
+):
+    """
+    For col-wise sharding of weight, we need to aggregate the
+    norm across all ranks before we can perform the proper re-norm.
+
+    Args:
+        max_norm: If given, each embedding vector with norm larger
+            than max_norm is renormalized to have norm max_norm.
+            Note: this will modify weight in-place.
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/_common.py:329]: Why would this modify weight in-place?

File: torch/distributed/_sharded_tensor/ops/embedding.py [hunk 2]
```diff
@@ -141,11 +200,25 @@ def _handle_col_wise_sharding(input, world_size, weight, local_shard, pg):
...
         pg: process group.
 
     Returns: final result of lookup.
     """
-    return _handle_col_wise_sharding_base(
+    if max_norm is not None:
+        # max_norm changes the weight in-place
+        local_shard = _handle_max_norm_col_wise(
+            max_norm, norm_type, local_shard, input, pg
+        )
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/embedding.py:219]: It seems like we are applying max_norm to the entire shard which doesn't seem right. The max_norm logic should be only applied to the embedding indices that we are looking up and not the whole shard.

File: torch/distributed/_sharded_tensor/ops/embedding_bag.py [hunk 3]
```diff
@@ -331,12 +432,21 @@ def _handle_row_wise_sharding(
             if gathered_per_sample_weights is not None
             else None
         )
+        # If input is None, passing in max_norm causes
+        # errors in CUDA.
+        if max_norm is not None and inp.size(0) == 0:
+            max_norm = None
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/embedding_bag.py:438]: What is the behavior for `nn.EmbeddingBag` when this happens? We should replicate that behavior here, silently setting `max_norm=None` doesn't seem right.

File: torch/distributed/_sharded_tensor/ops/embedding_bag.py [hunk 4]
```diff
@@ -367,27 +485,22 @@ def _handle_row_wise_sharding(
     # For Mean, we cannot do the division until very end because the sum of means
     # not equal to the mean of sum. (Divisor is different)
     if mode == "mean":
-        # For a 2D tensor, we just average by col number
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/embedding_bag.py:370]: Do we no longer need this handling for 2D tensor?

**Review by pritamdamania87 — COMMENTED** [review]
Awesome work!

File: torch/distributed/_sharded_tensor/ops/_common.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/_common.py:329]: Oh, it's required by the wiki.

File: torch/distributed/_sharded_tensor/ops/embedding.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/embedding.py:219]: Yes... I thought about this. If we want to copy the exact behavior, we could end up adding more communication overhead because we need to first let each rank knows which IDs are being looked up. If you feel this is necessary, I can modify it.

File: torch/distributed/_sharded_tensor/ops/embedding_bag.py (hunk 3 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/embedding_bag.py:438]: Oh, this is fine because the input here is None. Interestingly, it throws exception if we don't set it as None. And for embedding bag, it does not raise exception. (It took me some time to learn this difference). So do you have any suggestion on how we should handle this?

File: torch/distributed/_sharded_tensor/ops/embedding_bag.py (hunk 4 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/embedding_bag.py:370]: Oh because I now merge the logic of 1D and 2D so that we don't need to treat them separately in the end.  So padding makes a 2D matrix similar like 1D because we can not directly use the col num as the divisor. So we created a new variable `split_sizes_1d_with_padding` to be the divisor.

**Comment by fduwjj** [comment]
Only renorm the indices of weight which are in the input.

**Comment by fduwjj** [comment]
Add the test for weight comparison so that we can make sure that it's working as expected.

File: torch/distributed/_sharded_tensor/ops/embedding.py (hunk 2 above)
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/ops/embedding.py]: There is additional communication overhead but reduced compute overhead since you update only a few indices. But in general isn't this a correctness issue? In the unit tests can you compare the weights after the max_norm update? I think vanilla nn.embeddingbag only renormalizes the indices you look up and not the whole embedding bag.

File: torch/distributed/_sharded_tensor/ops/embedding_bag.py (hunk 3 above)
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/ops/embedding_bag.py:440]: Ah so if input is empty and max_norm is not None, the embedding_bag lookup will throw an exception? Btw, if input is empty do we even need to do the embedding lookup?

File: test/distributed/_sharded_tensor/ops/test_embedding.py [hunk 5]
```diff
@@ -51,20 +76,52 @@ def _run_sharded_embedding(self, spec, input_size, num_embeddings, embedding_dim
...
+            dist.all_gather(gathered_inputs, inp)
+            unique_inp = torch.unique(torch.cat(gathered_inputs))
+            local_embedding(unique_inp)
+
         # Run local computation
         local_output = local_embedding(inp)
 
+        # Compare local weight and shared one to ensure the renorm
+        # as expected.
+        if max_norm is not None:
```
Comment (pritamdamania87) [inline test/distributed/_sharded_tensor/ops/test_embedding.py:92]: Are we testing cases here where the inputs to the embedding table are a subset of the total embeddings? If so, I'm wondering why this test doesn't fail since in sharded case update the entire embedding table but in local we only update the looked up indices. Unless I'm missing something here.

**Review by pritamdamania87 — COMMENTED** [review]
Mostly have a concern around the max_norm logic, let me know if it makes sense.

File: test/distributed/_sharded_tensor/ops/test_embedding.py (hunk 5 above)
Comment (fduwjj) [reply test/distributed/_sharded_tensor/ops/test_embedding.py:92]: To answer your question:

1. If we don't compare the weight, we will not know the difference because the results are not aware of whether we do renorm globally or to the looked up indices only. Because if we don't look up these indices, the weight change for these indices are not tested. 
2. I gather all inputs from all ranks and send them to each local first, so that shared weights are the same as the local one. And our test cases do cover that. If let's say we don't sync the input across all ranks, the test does fail.


Does this answer your question?

File: torch/distributed/_sharded_tensor/ops/embedding.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/embedding.py]: Yes, you are right. I did two things here:
1. I did the gather of input in this stage so that, no extra communication cost and we now only renorm the indices we look up. So this is correct now. I should have mentioned clearly in the update comment.
2. The unit test does compare the weight now. The only thing is that I need to first apply inputs from all ranks to the embedding/embedding bag so that the local will be renorm as sharded. Otherwise, the test would fai.

File: test/distributed/_sharded_tensor/ops/test_embedding.py [hunk 6]
```diff
@@ -51,20 +76,52 @@ def _run_sharded_embedding(self, spec, input_size, num_embeddings, embedding_dim
         inp = torch.randint(0, num_embeddings, tuple(input_size)).cuda(self.rank)
         sharded_output = sharded_embedding(inp)
 
+        # If max_norm is set, we need to ensure that the renorm has been applied across
+        # inputs from all ranks.
+        if max_norm is not None:
+            gathered_inputs = [torch.zeros_like(inp) for _ in range(TEST_GPU_NUM)]
+            dist.all_gather(gathered_inputs, inp)
+            unique_inp = torch.unique(torch.cat(gathered_inputs))
+            local_embedding(unique_inp)
```
Comment (fduwjj) [inline test/distributed/_sharded_tensor/ops/test_embedding.py:85]: So here, I apply the inputs from all ranks to the local embedding. Otherwise the test would fail. cc: @pritamdamania87

File: torch/distributed/_sharded_tensor/ops/_common.py [hunk 7]
```diff
@@ -268,3 +314,65 @@ def _communicate_list_to_each_rank(
...
+
+    Return:
+        local_shard_norm_renormed: local_shard re-normed to max_norm if the norm is larger
+            than it.
+        gathered_inputs: list of inputs from all ranks.
+    """
+    norm_type = norm_type if norm_type is not None else 2.0
+    # allgather the inputs first.
+    gathered_inputs = [torch.zeros_like(input) for _ in range(world_size)]
+    dist.all_gather(gathered_inputs, input, group=pg)
```
Comment (fduwjj) [inline torch/distributed/_sharded_tensor/ops/_common.py:351]: So here we do the gather so we can skip the gather later-on. This saves us some communication overhead.

File: test/distributed/_sharded_tensor/ops/test_embedding.py [hunk 8]
```diff
@@ -109,11 +132,13 @@ def test_sharded_embedding_colwise(self):
             self._run_sharded_embedding(spec, [4], 15, 14)
             self._run_sharded_embedding(spec, [34], 15, 14, padding_idx=10)
             self._run_sharded_embedding(spec, [8, 6, 5, 4], 23, 13, padding_idx=12)
-            self._run_sharded_embedding(spec, [4, 5, 6], 23, 13, max_norm=2.5)
```
Comment (pritamdamania87) [inline test/distributed/_sharded_tensor/ops/test_embedding.py]: Why was this test removed?

File: torch/distributed/_sharded_tensor/ops/_common.py [hunk 9]
```diff
@@ -268,3 +314,65 @@ def _communicate_list_to_each_rank(
...
+    normalized_tensor = torch.where(
+        local_shard_norm > max_norm_tensor, max_norm_tensor, local_shard_norm
+    )
+    # Make sure divisor is not zero.
+    local_shard_norm[local_shard_norm == 0.0] = 1.0
+    local_shard_norm_renormed = (
+        torch.div(torch.mul(local_shard_t, normalized_tensor), local_shard_norm)
+        .t()
+        .contiguous()
+    )
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/_common.py:377]: My major confusion is with this logic. It seems like we are applying renorm to the entire local_shard and not only to rows of local_shard that have been queried. This doesn't seem to be correct and I'm not sure why the unit tests are not catching this issue and failing.

Unless I'm missing something here.

File: test/distributed/_sharded_tensor/ops/test_embedding.py (hunk 8 above)
Comment (fduwjj) [reply test/distributed/_sharded_tensor/ops/test_embedding.py]: Opps, might be because of the merge conflict.

**Review by pritamdamania87 — APPROVED** [review]

**Comment by fduwjj** [comment]
Resolved the question offline.
